conceptnet_sme/semantic_matching.py

```diff
@@ -519,18 +537,322 @@ def export(self, dirname):
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            weight_decay = group['weight_decay']
+            momentum = group['momentum']
+            dampening = group['dampening']
+            nesterov = group['nesterov']
+
+            for p in group['params']:
+                if p.grad is None:
+                    continue
+                d_p = p.grad.data
+                d_p_is_sparse = (d_p.layout == torch.sparse_coo)
+                if weight_decay != 0:
+                    if not d_p_is_sparse:
+                        d_p.add_(weight_decay, p.data)
+                    else:
+                        # Update the gradient via weight decay only at
+                        # locations where it is already non-zero; this
+                        # isn't quite the same as actual weight decay but
+                        # it avoids ballooning the memory consumption.
+                        d_p = d_p.coalesce()
+                        p.grad.data = d_p
+                        if len(d_p._indices().size()) < 2: # empty sparse tensor
+                            pass
+                        else:
+                            # Find the indices of the nonzero entries in the
+                            # gradient, assuming row-major storage order, by
+                            # finding the index of each nonzero subtensor's
+                            # first entry and adding offsets to the other
+                            # entries.
+                            assert p.data.size() == d_p.size()
+                            n_sparse_dims, n_values = d_p._indices().size()
+                            values_shape = tuple(d_p._values().size())
+                            assert len(values_shape) > 0
+                            assert n_values == values_shape[0]
+                            value_size = int(np.prod(np.array(
+                                values_shape[1:])))
+                            factors = np.array(d_p.size())
+                            factors = np.flip(factors, axis=0)
+                            factors = np.append(np.array([1]), factors[:-1])
+                            factors = np.cumprod(factors)
+                            factors = np.flip(factors, axis=0)
+                            factors = factors.tolist() # lose negative strides
+                            assert n_sparse_dims <= len(factors)
+                            flat_indices0 = torch.zeros(
+                                n_values, dtype=torch.int64, device=d_p.device)
+                            for i_sparse_dim in range(n_sparse_dims):
+                                flat_indices0.add_(
+                                    factors[i_sparse_dim],
+                                    d_p._indices()[i_sparse_dim])
+                            flat_indices0.unsqueeze_(1)
+                            offsets = torch.arange(value_size, 
```

Comment:
The code gets very hard to read around here. It's okay to use more than 80 columns (up to 100) when it actually aids readability.

Reply:
I've been bitten in the past by tools that failed in various ways with long source lines, but I'll try to honor your preference on this going forward.